flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureRateRestartBackoffTimeStrategy.java

```diff
@@ -0,0 +1,156 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.executiongraph.failover.flip1;
+
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.util.clock.Clock;
+import org.apache.flink.runtime.util.clock.SystemClock;
+
+import scala.concurrent.duration.Duration;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import static org.apache.flink.configuration.RestartBackoffTimeStrategyOptions.RESTART_BACKOFF_TIME_STRATEGY_FAILURE_RATE_FAILURE_RATE_BACKOFF_TIME;
+import static org.apache.flink.configuration.RestartBackoffTimeStrategyOptions.RESTART_BACKOFF_TIME_STRATEGY_FAILURE_RATE_FAILURE_RATE_INTERVAL;
+import static org.apache.flink.configuration.RestartBackoffTimeStrategyOptions.RESTART_BACKOFF_TIME_STRATEGY_FAILURE_RATE_MAX_FAILURES_PER_INTERVAL;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Restart strategy which can restart when failure rate is not exceeded.
+ */
+public class FailureRateRestartBackoffTimeStrategy implements RestartBackoffTimeStrategy {
+
+	private final long failuresIntervalMS;
+
+	private final long backoffTimeMS;
+
+	private final int maxFailuresPerInterval;
+
+	private final Deque<Long> failureTimestamps;
+
+	private final String strategyString;
+
+	private final Clock clock;
+
+	FailureRateRestartBackoffTimeStrategy(Clock clock, int maxFailuresPerInterval, long failuresIntervalMS, long backoffTimeMS) {
+
+		checkArgument(maxFailuresPerInterval > 0, "Maximum number of restart attempts per time unit must be greater than 0.");
+		checkArgument(failuresIntervalMS > 0, "Failures interval must be greater than 0 ms.");
+		checkArgument(backoffTimeMS >= 0, "Backoff time must be at least 0 ms.");
+
+		this.failuresIntervalMS = failuresIntervalMS;
+		this.backoffTimeMS = backoffTimeMS;
+		this.maxFailuresPerInterval = maxFailuresPerInterval;
+		this.failureTimestamps = new ArrayDeque<>(maxFailuresPerInterval);
+		this.strategyString = generateStrategyString();
+		this.clock = checkNotNull(clock);
+	}
+
+	@Override
+	public boolean canRestart() {
+		if (isFailureTimestampsQueueFull()) {
+			Long now = clock.absoluteTimeMillis();
+			Long earliestFailure = failureTimestamps.peek();
+
+			return (now - earliestFailure) > failuresIntervalMS;
```

Comment:
How exactly is the interval defined? Assuming my interval is `2L` and I have two failures `T(0), T(2)` occurring at time `0` and `2`, then it would not restart. Thus, the boundaries of the interval `[0, 2]` are inclusive. An alternative could be to make the right boundary exclusive `[0, 2)`.

Reply:
Yes, I think both are OK. Let’s keep it as now, since it is also consistent with the definition of legacy `FailureRateRestartStrategy`.